### ISSUE TITLE
plugin/kubernetes: ttl for different service type

### ISSUE THREAD

**Issue by ckcd** [coredns/coredns#2518]
Hi,

I have a question about `ttl` of services with different type:

In Kubernetes, dns `A` record of a `clusterIP service` is the IP address itself, but dns `A` record of a `headless service` is a list of endpoints addresses.

In my rough opinion, the `clusterIP service` is stable and we won't change it frequently, and it changes only when the service be deleted.  and it's IP address is stable which means the dns record will not change.

Meanwhile, the dns record of `headless service` is unstable because the `endpoints` will change when `pod` changes, and in k8s `pod` will be frequent kill and start so it's ip address has changed.

So i'm thinking about set different `ttl` for these different services. which may be present as:
1. for default ttl, give `clusterIP service` a long ttl (maybe 60s?) and give `headless service` a short ttl (maybe 5s?)
2. give more parameter of `plugin/kubernetes` in `Corefile`, so users can defines different ttl for different services.

I'm not sure whether it's necessary, please correct me if i have some misunderstanding.

Thanks !!!

**Comment by chrisohaver** [comment]
>I'm not sure whether it's necessary

TL;DR: I think it could benefit node local dns cache deployments.

In CoreDNS, the _kubernetes_ plugin keeps its own cache, which is updated in "real time" asynchronously via a watch on the Kubernetes API.  So, even if you deploy _kubernetes_ plugin with a TTL of 0, or if you deploy CoreDNS with no _cache_ plugin at all, queries for Kubernetes records do not cause round trips to the K8s api per request.  In other words, there is not a significant performance difference when pulling a record from the _cache_ plugin vs the _kubernetes_ plugin.  So, from a coredns performance perspective, the TTL on _kubernetes_ records doesn't really factor in strongly.

But TTL would come into play with clients that might pay attention to the TTL, or if the the client reaching coredns through a local caching DNS, such as would happen with node local dns caching in k8s (which i think is becoming more common and may become standard soon).  In a node local dns caching deployment, a higher TTL for cluster ip services would reduce DNS traffic to the cluster DNS.

> give more parameter of `plugin/kubernetes` in Corefile, so users can defines different ttl for different services.

I think we could do this by extending the existing `ttl` directive, allowing an optional second parameter.
The second parameter if specified being the TTL to use for Cluster IP services, the first used for Pods, Endpoints, headless Services, and negative responses.

**Comment by ckcd** [comment]
@chrisohaver 
Thanks for your reply !

> In other words, there is not a significant performance difference when pulling a record from the cache plugin vs the kubernetes plugin.

Agree with you. Actually at the beginning i have this idea in KubeDNS because it causes multiple dnsmasq --> skydns(work for services) which are two different containers, and it may affect performance. But in CoreDNS there is just one container and different `plugin` can communicate directly, so maybe the performance is OK.

> or if the the client reaching coredns through a local caching DNS, such as would happen with node local dns caching in k8s.

Yes, actually in my cluster, i gave some important pods a `sidecar dns cache` which is similar with node local dns caching.

> I think we could do this by extending the existing ttl directive, allowing an optional second parameter.

Good idea, i will take a look of this method and work for it.

**Comment by miekg** [comment]
This needs work in the DNS spec to make it happen.

**Comment by miekg** [comment]
If we have a spec change, this can be reopened.

**Comment by frimik** [comment]
> I think we could do this by extending the existing `ttl` directive, allowing an optional second parameter.
> The second parameter if specified being the TTL to use for Cluster IP services, the first used for Pods, Endpoints, headless Services, and negative responses.

This was *exactly* what I was looking for today... to have a much longer TTL on the Pod A records, which never change...
